Add `--only-root` option to install command, excluding all dependencies.

chadac opened the pull request:
Per the related issue, added the `--only-root` option to the `InstallCommand` (as hinted in the issue) to exclude all dependency groups from install.

This is already [expected behavior on the part of the installer](https://github.com/python-poetry/poetry/blob/master/tests/installation/test_installer.py#L383) so no additional tests are needed to ensure this works with an empty list of dependency groups.

# Pull Request Check List

Resolves: #3377

<!-- This is just a reminder about the most common mistakes. Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://python-poetry.org/docs/contributing/) at least once, it will save you unnecessary review cycles! -->

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

<!-- If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing! -->

radoering commented:
Out of curiosity: What happens if I run `poetry install --only-root --no-root`?

chadac commented:
> Out of curiosity: What happens if I run `poetry install --only-root --no-root`?

Haha, good question :) I tried it out, looks like it's effectively a noop minus creating the lockfile if it isn't present.

neersighted commented:
> > Out of curiosity: What happens if I run `poetry install --only-root --no-root`?
> 
> Haha, good question :) I tried it out, looks like it's effectively a noop minus creating the lockfile if it isn't present.

I'd like to see a meaningful message for the user when this case happens, e.g. https://github.com/python-poetry/poetry/blob/6d01f074f10ce08af33ce59aa0c5065dac24244d/src/poetry/console/commands/install.py#L77

abn commented:
Is this something that should apply to all group enabled commands? For example, this won't make much sense for update or export.

chadac commented:
> > > Out of curiosity: What happens if I run `poetry install --only-root --no-root`?
> > 
> > 
> > Haha, good question :) I tried it out, looks like it's effectively a noop minus creating the lockfile if it isn't present.
> 
> I'd like to see a meaningful message for the user when this case happens, e.g.
> 
> https://github.com/python-poetry/poetry/blob/6d01f074f10ce08af33ce59aa0c5065dac24244d/src/poetry/console/commands/install.py#L77-L83

Good point, I've gone ahead and added logic to catch this similar to that case, and set it up to exit when both options are present.

Should I also have this exit when both `--only-root` and `--with`/`--without`/`--only` are also passed? Right now it's currently a warning.



> Is this something that should apply to all group enabled commands? For example, this won't make much sense for update or export.

True, I've gone ahead and migrated it into `install.py` and modified the logic accordingly.

neersighted commented on src/poetry/console/commands/install.py:
```diff
@@ -82,6 +96,24 @@ def handle(self) -> int:
             return 1
 
+        if self.option("only-root") and any(
+            [self.option(key) for key in {"with", "without", "only"}]
```
```suggestion
            self.option(key) for key in {"with", "without", "only"}
```

You should be able to drop the brackets here and just let this become a generator expression instead of a list comprehension.

neersighted commented:
> Should I also have this exit when both `--only-root` and `--with`/`--without`/`--only` are also passed? Right now it's currently a warning.

I personally prefer an error as well, because an inattentive developer may munge a command line out of history. If they don't notice a warning message they may never understand the unexpected/inconsistent results they get until much later. I'd rather fail first as what the user is asking us to do is nonsensical and we should ask them to re-evaluate their request.

(this is all hypothetical of course, I would never make a silly mistake like re-using a command from history and combining options in a nonsensical manner :smile:)

chadac commented:
> I personally prefer an error as well, because an inattentive developer may munge a command line out of history. If they don't notice a warning message they may never understand the unexpected/inconsistent results they get until much later. I'd rather fail first as what the user is asking us to do is nonsensical and we should ask them to re-evaluate their request.
> 
> (this is all hypothetical of course, I would never make a silly mistake like re-using a command from history and combining options in a nonsensical manner smile)

Yeah, that's definitely true. Plus I imagine swapping warnings to errors in the future is a whole lot harder than the other way around... might need to keep that in mind for some of my own tools lol

neersighted commented on docs/cli.md:
````diff
@@ -167,6 +167,12 @@ It's also possible to only install specific dependency groups by using the `only
 poetry install --only test,docs
 ```
 
+To only install the project itself with no dependencies, use the `only-root` flag.
````
```suggestion
To only install the project itself with no dependencies, use the `--only-root` flag.
```

neersighted commented on docs/managing-dependencies.md:
```diff
@@ -158,6 +158,15 @@ poetry install --only main
 {{% /note %}}
 
+{{% note %}}
+If you only want to install the project root, and no other dependencies, you can use
```
```suggestion
If you want to install the project root, and no other dependencies, you can use
```
